### PR TITLE
Add `unexplained-noqa` rule

### PR DESCRIPTION
**Pull request by tjkuson**
## Summary

Add `unexplained-noqa` rule as `RUF101` that checks for `noqa` directives without comments. For example,

```python
foo == ""  # noqa: PLC1901
```

would be flagged and

```python
foo == ""  # noqa: PLC1901  # Check for empty string, not just falsiness.
```

would not.

Includes documentation. Related to #5182.

## Test Plan

`cargo test`

### CODE REVIEW

File: crates/ruff/src/checkers/noqa.rs [hunk 1]
```diff
@@ -27,12 +35,34 @@ pub(crate) fn check_noqa(
     // Extract all `noqa` directives.
     let mut noqa_directives = NoqaDirectives::from_commented_ranges(comment_ranges, locator);
 
+    if settings.rules.enabled(Rule::UnexplainedNOQA) {
+        for comment_range in comment_ranges {
+            let text: &str = &locator.contents()[*comment_range];
+            if let Some(caps) = NOQA_LINE_EXPLANATION_REGEX.captures(text) {
```
Comment (MichaReiser) [inline crates/ruff/src/checkers/noqa.rs:41]: Captures are kind of expensive. Can we use `text.trim_start().find("# noqa")` to find the starting position. Or @charliermarsh do we already have infrastructure to parse out noqa comments (we should, right?)

Comment (konstin) [reply crates/ruff/src/checkers/noqa.rs:41]: i think the regex is partially copied from https://github.com/astral-sh/ruff/blob/f401050878a3ca2506c65c29968a943557f1fdd8/crates/ruff/src/noqa.rs#L21-L26

Comment (tjkuson) [reply crates/ruff/src/checkers/noqa.rs:41]: Yes, I just adapted the current `noqa` regex as it seems to be how `noqa` directives are currently parsed. Unless there is a better way to parse the comments, I'll implement the suggestion to use `text.trim_start().find("# noqa")`.

**Comment by tjkuson** [comment]
Changed the example used in the documentation, as the previous example is now irrelevant due to #5264.

**Comment by tjkuson** [comment]
It doesn't work with codes that don't end in numbers, which is odd. I have made the PR a draft until this is fixed.

**Comment by tjkuson** [comment]
I will close it for now, then.

**Pull request closed without merging**